File: js/src/context.ts
type Agento11yContextValues = {
  conversationId?: string;
  conversationTitle?: string;
  userId?: string;
  agentName?: string;
  agentVersion?: string;
};

type ContextStorage<T> = {
  getStore(): T | undefined;
  run<R>(store: T, callback: () => R): R;
};

type AsyncLocalStorageConstructor = new <T>() => ContextStorage<T>;

export function withConversationId<T>(conversationId: string, callback: () => T): T {
  return runWithContext({ conversationId }, callback);
}

export function withConversationTitle<T>(conversationTitle: string, callback: () => T): T {
  return runWithContext({ conversationTitle }, callback);
}

export function withUserId<T>(userId: string, callback: () => T): T {
  return runWithContext({ userId }, callback);
}

export function withAgentName<T>(agentName: string, callback: () => T): T {
  return runWithContext({ agentName }, callback);
}

export function withAgentVersion<T>(agentVersion: string, callback: () => T): T {
  return runWithContext({ agentVersion }, callback);
}

export function conversationIdFromContext(): string | undefined {
  return normalizedString(storage.getStore()?.conversationId);
}

export function conversationTitleFromContext(): string | undefined {
  return normalizedString(storage.getStore()?.conversationTitle);
}

export function userIdFromContext(): string | undefined {
  return normalizedString(storage.getStore()?.userId);
}

export function agentNameFromContext(): string | undefined {
  return normalizedString(storage.getStore()?.agentName);
}

export function agentVersionFromContext(): string | undefined {
  return normalizedString(storage.getStore()?.agentVersion);
}

function runWithContext<T>(nextValues: Agento11yContextValues, callback: () => T): T {
  const currentValues = storage.getStore() ?? {};
  const mergedValues: Agento11yContextValues = { ...currentValues };

  for (const [key, value] of Object.entries(nextValues)) {
    const normalized = normalizedString(value);
    if (normalized === undefined) {
      delete mergedValues[key as keyof Agento11yContextValues];
      continue;
    }
    mergedValues[key as keyof Agento11yContextValues] = normalized;
  }

  return storage.run(mergedValues, callback);
}

function normalizedString(value: string | undefined): string | undefined {
  if (value === undefined) {
    return undefined;
  }
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function resolveNodeAsyncLocalStorage(): AsyncLocalStorageConstructor | undefined {
  const processWithBuiltins = (globalThis as { process?: { getBuiltinModule?: (id: string) => unknown } }).process;
  const module = processWithBuiltins?.getBuiltinModule?.('async_hooks') as
    | { AsyncLocalStorage?: AsyncLocalStorageConstructor }
    | undefined;
  return module?.AsyncLocalStorage;
}

// NoopContextStorage is used when the runtime does not expose Node's
// AsyncLocalStorage (for example some edge runtimes). A naive mutable-global
// fallback would silently mix contexts across concurrent async chains, which
// is worse than no propagation at all because telemetry would attribute
// records to the wrong conversation/user. Disabling propagation makes the
// limitation observable: callers must pass identifiers explicitly via the
// generation/tool start fields when running in such a runtime.
class NoopContextStorage<T> implements ContextStorage<T> {
  getStore(): T | undefined {
    return undefined;
  }

  run<R>(_store: T, callback: () => R): R {
    return callback();
  }
}

const AsyncLocalStorage = resolveNodeAsyncLocalStorage();
const storage: ContextStorage<Agento11yContextValues> =
  AsyncLocalStorage !== undefined
    ? new AsyncLocalStorage<Agento11yContextValues>()
    : new NoopContextStorage<Agento11yContextValues>();

if (AsyncLocalStorage === undefined && typeof console !== 'undefined' && typeof console.warn === 'function') {
  console.warn(
    'agento11y: AsyncLocalStorage is not available in this runtime; context helpers ' +
      '(withConversationId, withUserId, withAgentName, ...) are disabled. ' +
      'Pass identifiers explicitly on each startGeneration / startToolExecution call.',
  );
}

File: python/sigil_sdk/context.py
"""Context helpers for conversation and agent identity propagation."""

from __future__ import annotations

import contextvars
from collections.abc import Iterator
from contextlib import contextmanager

_conversation_id: contextvars.ContextVar[str | None] = contextvars.ContextVar("sigil_conversation_id", default=None)
_conversation_title: contextvars.ContextVar[str | None] = contextvars.ContextVar(
    "sigil_conversation_title", default=None
)
_user_id: contextvars.ContextVar[str | None] = contextvars.ContextVar("sigil_user_id", default=None)
_agent_name: contextvars.ContextVar[str | None] = contextvars.ContextVar("sigil_agent_name", default=None)
_agent_version: contextvars.ContextVar[str | None] = contextvars.ContextVar("sigil_agent_version", default=None)


@contextmanager
def with_conversation_id(conversation_id: str) -> Iterator[None]:
    """Sets conversation id within a context block."""

    token = _conversation_id.set(conversation_id)
    try:
        yield
    finally:
        _conversation_id.reset(token)


@contextmanager
def with_conversation_title(conversation_title: str) -> Iterator[None]:
    """Sets conversation title within a context block."""

    token = _conversation_title.set(conversation_title)
    try:
        yield
    finally:
        _conversation_title.reset(token)


@contextmanager
def with_user_id(user_id: str) -> Iterator[None]:
    """Sets user id within a context block."""

    token = _user_id.set(user_id)
    try:
        yield
    finally:
        _user_id.reset(token)


@contextmanager
def with_agent_name(agent_name: str) -> Iterator[None]:
    """Sets agent name within a context block."""

    token = _agent_name.set(agent_name)
    try:
        yield
    finally:
        _agent_name.reset(token)


@contextmanager
def with_agent_version(agent_version: str) -> Iterator[None]:
    """Sets agent version within a context block."""

    token = _agent_version.set(agent_version)
    try:
        yield
    finally:
        _agent_version.reset(token)


def conversation_id_from_context() -> str | None:
    """Returns the current conversation id from context variables."""

    return _conversation_id.get()


def agent_name_from_context() -> str | None:
    """Returns the current agent name from context variables."""

    return _agent_name.get()


def agent_version_from_context() -> str | None:
    """Returns the current agent version from context variables."""

    return _agent_version.get()


def conversation_title_from_context() -> str | None:
    """Returns the current conversation title from context variables."""

    return _conversation_title.get()


def user_id_from_context() -> str | None:
    """Returns the current user id from context variables."""

    return _user_id.get()
